core: Add support for ECDSA identities

davxy opened the pull request:
Closes https://github.com/libp2p/rust-libp2p/issues/556

Implementation compliant with https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#ecdsa

davxy committed:
First rough implementation of ECDSA support

davxy committed:
ECDSA keys support

mxinden reviewed (approved):
Thank you @davxy for the pull request.

This looks good to me, though I don't think I should be the only reviewer on a crypto focused pull request.

@dignifiedquire or @burdges do you have opinions on `p256` vs `ring`?

dignifiedquire reviewed (commented):
implementation from a crypto standpoint looks good to me

thomaseizinger commented on core/src/identity/ecdsa.rs:
```diff
@@ -0,0 +1,161 @@
+
+impl PublicKey {
+    /// Verify an ECDSA signature on a message using the public key.
+    pub fn verify(&self, msg: &[u8], sig: &[u8]) -> bool {
```
Might be good to annotate this with `#[must_use]` to avoid accidentially discarding the result.

davxy replied:
I've followed the design of the other key types. I can add the attribute, but then I suggest to do the same for the others.

thomaseizinger replied:
I didn't know it is missing from the others!

I think it would be good to add but we can probably leave this to a follow-up PR then.

thomaseizinger commented on core/src/identity/ecdsa.rs:
```diff
@@ -0,0 +1,161 @@
+
+/// An ECDSA secret key.
+#[derive(Clone)]
+pub struct SecretKey(SigningKey);
```
There is currently no way of constructing one of these from bytes. How are nodes meant to have to persistent identity if all we can do is generate random ones?

davxy replied:
Legitimate. I add the functions

thomaseizinger commented on core/src/identity/ecdsa.rs:
```diff
@@ -0,0 +1,161 @@
+
+impl fmt::Debug for SecretKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "SecretKey")
```
Over at rust-bitcoin, we settled on including a fingerprint (first couple of bytes of tagged hash) in the debug output to make it possible to differentiate different secret keys. Thoughts on doing the same thing here?

davxy replied:
I've followed the design of the other key types. I can add the attribute, but then I suggest to do the same for the others.

Can you elaborate on what should be used as private key fingerprint?

thomaseizinger replied:
> I've followed the design of the other key types. I can add the attribute, but then I suggest to do the same for the others.

Leave it then, would probably need more discussion overall :)

> Can you elaborate on what should be used as private key fingerprint?

I'd suggest using the first couple of bytes of a tagged hash (ie `sha256(b"ECDSA-secret-key" + bytes)[0..8]`) but if none of the other secret keys do this, we don't need to add it here.

thomaseizinger commented on core/src/identity/ecdsa.rs:
```diff
@@ -0,0 +1,161 @@
+    // Decode a public key from a DER encoded byte buffer as defined for SEC1 standard.
+    pub fn decode_der(k: &[u8]) -> Result<PublicKey, DecodingError> {
+        VerifyingKey::from_sec1_bytes(k)
+            .map_err(|_| DecodingError::new("failed to parse ecdsa p256 public key"))
```
Could `?` after this to avoid the last `.map`.

thomaseizinger replied:
We are dropping the source error here. Is that what we do for other keypairs as well? I think we should include the source error message.

davxy replied:
The second `map` has been inserted to map from `Result<VerifyingKey>` into a `Result<PublicKey>`.

With PublicKey defined as : `struct PublicKey(VerifyingKey)`

davxy replied:
This should have been resolved.

Where possible the errors are propagated using `source()`. 

There is an error that cannot be stored within the source because it doesn't implement `Error`. In this case I've added some details to the error message.

```
        let enc_pt = EncodedPoint::from_bytes(k).map_err(|_| {   // <- this doesn't implement error on `no_std` builds
            DecodingError::new("failed to parse ecdsa p256 public key, bad point encoding")
        })?;
```

thomaseizinger reviewed (commented):
Thanks for the contribution!

Some comments / discussion points.

davxy committed:
Fix ECDSA ASN.1 PublicKey serialization

Furthermore SecretKey can now be serialized/deserialized to/from a byte
vector.

thomaseizinger reviewed (approved):
Looks good to me, thank you!

mxinden committed:
Merge branch 'master' into ecdsa_identity

mxinden committed:
core/CHANGELOG: Add entry

mxinden reviewed (approved):
Thanks @davxy for the contribution.

Also tested with https://github.com/libp2p/rust-libp2p/blob/master/examples/ipfs-kad.rs using ECDSA identity.

mxinden committed:
core/src/identity/ecdsa: Fix clippy warning
